invenio_vocabularies/services/facets.py

```diff
@@ -19,6 +22,49 @@
 from ..proxies import current_service
 
 
+def get_service(service_name):
+    """Get the service object by name.
+
+    It is required to access the registry lazily, to avoid out of
+    application context errors.
+    """
+    return (
+        current_service_registry.get(service_name) if service_name else current_service
+    )
+
+
+def get_vocabs(service_name, type, ids, fields):
+    """Fetch vocabulary values by ids, using the service."""
+    service = get_service(service_name)
+    vocabs = service.read_many(AnonymousIdentity(), type=type, ids=ids, fields=fields)
+    return list(vocabs.hits)  # the service returns a generator
+
+
+def get_ttl_hash(seconds=3600):
+    """Return the same value withing `seconds` time period. Default 1 hour."""
+    return round(time.time() / seconds)
+
+
+@lru_cache(maxsize=32)
+def cached_vocabs(service_name, type, ids, fields, ttl_hash=None):
```

Comment:
question/shleve: from what I understand this caches results for a service/vocab-type on the level of `ids`, i.e. the calls with ids `["dataset", "software", "figure"]` and `["software"]` are stored separately?

I guess this is to keep things simple, otherwise caching on individual IDs and reconstructing the result gets a bit complicated.

Reply:
> question/shleve: from what I understand this caches results for a service/vocab-type on the level of `ids`, i.e. the calls with ids `["dataset", "software", "figure"]` and `["software"]` are stored separately?
> 
> I guess this is to keep things simple, otherwise caching on individual IDs and reconstructing the result gets a bit complicated.

Yes indeed. Would you prefer to cache individual ids?
The cache has been created having its usage in mind, where most of the time, a list of `ids` is passed. We could go for caching each `id` too, it might complexify things.
Do you see any particular advantage? The only advantage I can think of is when only one `id` is fetched. It will work anyway, but the first time you will have a `miss`.

Reply:
I think the list of IDs changes e.g. based on the search query you make, i.e. if I search records for `Python`, I will probably not get any `Text` resource type results in the facets.

Given that this is sort of a micro-caching strategy, on the level of individual uWSGI processes it should be fine for now (also it will get more complex to mangle the cached + fetched results). e.g. in Zenodo there's a big enough volume of records spread across all the different resource types, so the set of unique resource type IDs coming from the facet/aggregatio when you make a search query is likely to be the same.

Reply:
Yes I agree. It was indeed done like this for simplicity and to measure if we really gain performance.
I can spend more time now to improve it, caching by id and handling the cache reconciliation of each id when multiple ids are fetched.